### PR TITLE
general: support for most launchpad clients

### PR DESCRIPTION
**Pull request by tupakkatapa**
This pull request introduces support for the first additional Ethereum client, which will be Geth ([go-ethereum](https://github.com/ethereum/go-ethereum)).

EDIT: Also adding rest of the [launchpad](https://launchpad.ethereum.org/en/) clients here, these include:

- Execution clients
  - [Besu](https://launchpad.ethereum.org/en/besu)
  - [Erigon](https://launchpad.ethereum.org/en/erigon)
  - [Geth](https://launchpad.ethereum.org/en/geth)
  - [Nethermind](https://launchpad.ethereum.org/en/nethermind)
  
- Consensus clients
  - [Lighthouse](https://launchpad.ethereum.org/en/lighthouse)
  - [Nimbus](https://launchpad.ethereum.org/en/nimbus)
  - [Prysm](https://launchpad.ethereum.org/en/prysm)
  - [Teku](https://launchpad.ethereum.org/en/teku)
  
  EDIT: Except for Nimbus, for more information, please read the comments below.

### CODE REVIEW

**Comment by jhvst** [comment]
Do all of the clients open the same port? I'm quite surprised if erigons torrent port of 42069 (conjugation of 420 and 69) is actually adopted unanimously.

Can you try on the ponkila alpha that the client actually start with these definitions?

We might need to also take into account multiple different data directories if someone wants to run two clients, and the option to define some other network than mainnet.

Can ethereum.nix options be programmatically reused?

**Comment by tupakkatapa** [comment]
> @jhvst

I overlooked the port configurations, I will check on that.

I have been testing these with QEMU and SSH'ing into it to check that everything works. Currently, there are no failed systemd units with any of the clients, but I believe that only Besu becomes active. This could be due to a lack of configuration or missing elements, so there might be some debugging and testing to do. 

Yes, I could also test them with bare-metal at `ponkila-alpha`, however, you need to give me access to the WireGuard configuration of that machine so I can still connect to it after booting into the configuration I am about to test. You could just drop it into `/var/mnt/secrets` and I'll take it from there.

I am not sure what you mean by taking multiple different data directories into account. It is already possible, like this:
```json
{
  "execution": {
    "erigon": {
      "enable": true,
      "dataDir": "/var/mnt/erigon"
    }
  },
  "consensus": {
    "lighthouse": {
      "enable": true,
      "dataDir": "/var/mnt/lighthouse"
    }
  }
}
```
If you are referring to running multiple instances of the same client, this is not possible by other means atm.

The current configuration is using ethereum.nix as an overlay, so we are currently only getting the packages. I have been considering the idea of getting another module inside of another, so modules would have recursive modularity. I can look into this if you want, we would be saved from a lot of work by doing this but we would lose a lot of control and congruity between the options.

**Comment by tupakkatapa** [comment]
Some `systemctl status` outputs from all execution clients at QEMU
```shell
core@homestaker ~ [3]> systemctl --failed
  UNIT LOAD ACTIVE SUB DESCRIPTION
0 loaded units listed.

core@homestaker ~> systemctl status erigon --no-pager
● erigon.service - execution, mainnet
     Loaded: loaded (/etc/systemd/system/erigon.service; enabled; preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Tue 2023-08-01 20:34:30 UTC; 2s ago
    Process: 2727 ExecStart=/nix/store/kdzlkyd1f4n8qlfw7mj1rdirmabyyl5i-unit-script-erigon-start/bin/erigon-start (code=exited, status=132)
   Main PID: 2727 (code=exited, status=132)
         IP: 0B in, 0B out
        CPU: 58ms

core@homestaker ~ [3]> systemctl status go-ethereum --no-pager
● go-ethereum.service - execution, mainnet
     Loaded: loaded (/etc/systemd/system/go-ethereum.service; enabled; preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Tue 2023-08-01 20:34:35 UTC; 1s ago
    Process: 2740 ExecStart=/nix/store/jhgmwap2i9kgwhgb7cw42bky8fmykrpp-unit-script-go-ethereum-start/bin/go-ethereum-start (code=exited, status=127)
   Main PID: 2740 (code=exited, status=127)
         IP: 0B in, 0B out
        CPU: 42ms

core@homestaker ~ [3]> systemctl status nethermind --no-pager
● nethermind.service - execution, mainnet
     Loaded: loaded (/etc/systemd/system/nethermind.service; enabled; preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Tue 2023-08-01 20:34:41 UTC; 3s ago
    Process: 2753 ExecStart=/nix/store/frhv8wrf2ahf47rxs80axbpk04slm2dh-unit-script-nethermind-start/bin/nethermind-start (code=exited, status=127)
   Main PID: 2753 (code=exited, status=127)
         IP: 0B in, 0B out
        CPU: 35ms

core@homestaker ~ [3]> systemctl status besu --no-pager
● besu.service - execution, mainnet
     Loaded: loaded (/etc/systemd/system/besu.service; enabled; preset: enabled)
     Active: active (running) since Tue 2023-08-01 20:34:20 UTC; 26s ago
   Main PID: 2681 (besu-start)
         IP: 0B in, 0B out
         IO: 0B read, 0B written
      Tasks: 15 (limit: 7859)
     Memory: 107.7M
        CPU: 23.286s
     CGroup: /system.slice/besu.service
             ├─2681 /nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash /nix/store/nfq3mll8234ks7vfw8d2n475155axgd9-unit-script-besu-start/bin/besu-start
             └─2682 /nix/store/14h03wdjaigk52ff7s49dffvchmj6fww-openjdk-19.0.2+7/bin/java -Dvertx.disableFileCPResolving=true -Dbesu.home=/nix/store/h1ym6awfm2rjdgy62jnr3nz3v2n112wi-besu-23.4.1 -Dlog4…

Aug 01 20:34:20 homestaker systemd[1]: Started execution, mainnet.
```

**Comment by tupakkatapa** [comment]
Nimbus fails to compile for some reason. Here are the outputs:

<details>

<summary>echo "$json_data" | nix run .#buidl -- -b homestakeros</summary>

```log
error: build of '/nix/store/8y60jq9j8bfywa5x18w3wd5xwv41xwip-nimbus-eth2-23.6.1.drv' on 'ssh://kari@buidl0.ponkila.com' failed: builder for '/nix/store/8y60jq9j8bfywa5x18w3wd5xwv41xwip-nimbus-eth2-23.6.1.drv' failed with exit code 2;
       last 10 log lines:
       >     setProperty(record_12012493182[1], "restUrl", restUrl)
       >     logAllDynamicProperties(
       >       defaultChroniclesStream, record_12012493182)
       >     flushRecord(record_12012493182[0])
       >     flushRecord(record_12012493182[1])
       > except CatchableError as err`gensym704:
       >   logLoggingFailure(cstring("Attempt to download genesis state timed out"),
       >                     err`gensym704)
       > nil' but expected 'ref ForkedHashedBeaconState'
       > make: *** [Makefile:423: nimbus_beacon_node] Error 1
       For full logs, run 'nix log /nix/store/8y60jq9j8bfywa5x18w3wd5xwv41xwip-nimbus-eth2-23.6.1.drv'.
error: builder for '/nix/store/8y60jq9j8bfywa5x18w3wd5xwv41xwip-nimbus-eth2-23.6.1.drv' failed with exit code 1;
       last 10 log lines:
       >     setProperty(record_12012493182[1], "restUrl", restUrl)
       >     logAllDynamicProperties(
       >       defaultChroniclesStream, record_12012493182)
       >     flushRecord(record_12012493182[0])
       >     flushRecord(record_12012493182[1])
       > except CatchableError as err`gensym704:
       >   logLoggingFailure(cstring("Attempt to download genesis state timed out"),
       >                     err`gensym704)
       > nil' but expected 'ref ForkedHashedBeaconState'
       > make: *** [Makefile:423: nimbus_beacon_node] Error 1
       For full logs, run 'nix log /nix/store/8y60jq9j8bfywa5x18w3wd5xwv41xwip-nimbus-eth2-23.6.1.drv'.
error: 1 dependencies of derivation '/nix/store/sccxwvgv4j7s4ws0pvfpl5r2mra510a3-man-paths.drv' failed to build
error: 1 dependencies of derivation '/nix/store/yp0igyn26l6hslsyawz0zzh6ca745l77-nimbus-eth2-23.6.1_fish-completions.drv' failed to build
error: 1 dependencies of derivation '/nix/store/naxij1j88pkvgqdl003wygfy3bgr75df-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/k46xj9048kzjw94mf5lz8m9dallyq1v4-unit-script-nimbus-start.drv' failed to build
error: 1 dependencies of derivation '/nix/store/iym8n6n3pfaf7clqpjgvdzil8fjgqfm5-nixos-system-homestaker-23.11.20230722.12303c6.drv' failed to build
error: 1 dependencies of derivation '/nix/store/67ynsr2vn1pc0wrjmca2a3jxpkxvc2ma-closure-info.drv' failed to build
error: 1 dependencies of derivation '/nix/store/cj207cl3qhc36w5m8jqvkmzsd3f7k19x-kexec-boot.drv' failed to build
error: 1 dependencies of derivation '/nix/store/7i14wn6lq6kzc4x1j6j0xdjfnv7g7yay-kexec-tree.drv' failed to build
```

</details>

<details>

<summary>nix log /nix/store/8y60jq9j8bfywa5x18w3wd5xwv41xwip-nimbus-eth2-23.6.1.drv</summary>

```log
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/8y60jq9j8bfywa5x18w3wd5xwv41xwip-nimbus-eth2-23.6.1.drv^*'
unpacking sources
unpacking source archive /nix/store/ylhbcvnl0ijnkah3srs0yc0ihwc490j4-source
source root is source
patching sources
updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./vendor/nim-libbacktrace/vendor/libbacktrace-upstream/config.sub
Updating Autotools / GNU config script to a newer upstream version: ./vendor/nim-libbacktrace/vendor/libbacktrace-upstream/config.guess
building
patching script interpreter paths in scripts vendor/nimbus-build-system/scripts
scripts/time_module_builds.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/start_nimbus_el_nodes.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/start_geth_nodes.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/signers/web3signer.sh: interpreter directive changed from "#!/usr/bin/bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/signers/nimbus.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/signers/custom.sh: interpreter directive changed from "#!/usr/bin/bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/setup_scenarios.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/run-nimbus.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/run-nimbus-in-devnet.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/run-geth-in-devnet.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/run-beacon-node.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/rotatelogs-compress.sh: interpreter directive changed from "#!/bin/sh" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/sh"
scripts/makedir.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/make_prometheus_config.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/make_packages.sh: interpreter directive changed from "#!/bin/bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/make_dist.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/launch_local_testnet.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/find_unchanged_consensus_spec_files.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/copyfile.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/compile_nim_program.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
scripts/clean-devnet-dir.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
vendor/nimbus-build-system/scripts/nimble.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
vendor/nimbus-build-system/scripts/git_show_deps.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bas>
vendor/nimbus-build-system/scripts/env.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
vendor/nimbus-build-system/scripts/create_nimble_link.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bi>
vendor/nimbus-build-system/scripts/create_nbs_paths.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/>
vendor/nimbus-build-system/scripts/build_rocksdb.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bas>
vendor/nimbus-build-system/scripts/build_nim.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash"
vendor/nimbus-build-system/scripts/add_submodule.sh: interpreter directive changed from "#!/usr/bin/env bash" to "/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bas>
Creating nimbus-build-system.paths
build flags: -j20 SHELL=/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/bash USE_SYSTEM_NIM=1 nimbus_beacon_node
23.6.1
Building: install/usr/lib/libbacktrace.a
Building: install/usr/lib/libbacktracenim.a
Building: build/generate_makefile
[using system Nim: /nix/store/ks51zgx7i3v7rv381c4k9ap8m9mqc044-x86_64-unknown-linux-gnu-nim-wrapper-1.6.14/bin/nim]
Build completed successfully: build/generate_makefile
Building: build/nimbus_beacon_node
[using system Nim: /nix/store/ks51zgx7i3v7rv381c4k9ap8m9mqc044-x86_64-unknown-linux-gnu-nim-wrapper-1.6.14/bin/nim]
Building: build/generate_makefile
[using system Nim: /nix/store/ks51zgx7i3v7rv381c4k9ap8m9mqc044-x86_64-unknown-linux-gnu-nim-wrapper-1.6.14/bin/nim]
Build completed successfully: build/generate_makefile
Building: build/nimbus_beacon_node
[using system Nim: /nix/store/ks51zgx7i3v7rv381c4k9ap8m9mqc044-x86_64-unknown-linux-gnu-nim-wrapper-1.6.14/bin/nim]
BLST: not using SSSE3
/build/source/beacon_chain/spec/datatypes/base.nim:653:15 template/generic instantiation of `toHex` from here
/nix/store/d4ivxmqa5n942kpq863ldc0zwh1gjfbc-nim-unwrapped-1.6.14/nim/lib/pure/strutils.nim:934:13 Warning: target type is larger than source type [CastSizes]
/build/source/ncli/e2store.nim:153:11 Warning: target type is larger than source type [CastSizes]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:933:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:935:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:938:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:939:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:933:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:935:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:938:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:939:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:933:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:935:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:938:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:905:64 template/generic instantiation of `async` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:921:12 template/generic instantiation of `withState` from here
/build/source/beacon_chain/gossip_processing/gossip_validation.nim:939:5 Warning: unreachable code after 'return' statement or '{.noReturn.}' proc [UnreachableCode]
/build/source/beacon_chain/trusted_node_sync.nim:79:56 template/generic instantiation of `async` from here
/build/source/beacon_chain/trusted_node_sync.nim:149:29 template/generic instantiation of `awaitWithTimeout` from here
/build/source/vendor/nim-eth/eth/async_utils.nim:16:5 Error: type mismatch: got 'untyped' for '
chronosInternalRetFuture.internalChild = cancelAndWait(f`gensym675)
yield chronosInternalRetFuture.internalChild
if chronosInternalRetFuture.internalMustCancel:
  raise (ref CancelledError)(msg: "Future operation cancelled!")
internalCheckComplete(chronosInternalRetFuture.internalChild)
bind logIMPL, bindSym, brForceOpen
try:
  block chroniclesLogStmt:
    if not topicsMatch(LogLevel(3), []):
      break chroniclesLogStmt
    var record_12012493182:
      defaultChroniclesStreamLogRecord
    mixin activateOutput
    record_12012493182 = default(typeof(record_12012493182))
    {.gcsafe.}:
      addr defaultChroniclesStreamOutputs
    [][0].currentRecordLevel = LogLevel(3)
    {.gcsafe.}:
      addr defaultChroniclesStreamOutputs
    [][1].currentRecordLevel = LogLevel(3)
    initLogRecord(record_12012493182[0], LogLevel(3), "",
                  "Attempt to download genesis state timed out")
    initLogRecord(record_12012493182[1], LogLevel(3), "",
                  "Attempt to download genesis state timed out")
    mixin setProperty, formatItIMPL
    setProperty(record_12012493182[0], "syncTarget", shortLog(syncTarget))
    setProperty(record_12012493182[1], "syncTarget", shortLog(syncTarget))
    mixin setProperty, formatItIMPL
    setProperty(record_12012493182[0], "restUrl", restUrl)
    setProperty(record_12012493182[1], "restUrl", restUrl)
    logAllDynamicProperties(
      defaultChroniclesStream, record_12012493182)
    flushRecord(record_12012493182[0])
    flushRecord(record_12012493182[1])
except CatchableError as err`gensym704:
  logLoggingFailure(cstring("Attempt to download genesis state timed out"),
                    err`gensym704)
nil' but expected 'ref ForkedHashedBeaconState'
make: *** [Makefile:423: nimbus_beacon_node] Error 1
```

</details>

**Comment by jhvst** [comment]
> ponkila-alpha

Sorry, meant `[ponkila-ephemeral-beta](https://github.com/ponkila/homestaking-infra/tree/main/nixosConfigurations/ponkila-ephemeral-beta)`. 

> If you are referring to running multiple instances of the same client, this is not possible by other means atm.

There could be an issue created of this, so we can deal with it later. It's quite common to have testnet and mainnet client be the same thing.

>The current configuration is using ethereum.nix as an overlay, so we are currently only getting the packages. I have been considering the idea of getting another module inside of another, so modules would have recursive modularity. I can look into this if you want, we would be saved from a lot of work by doing this but we would lose a lot of control and congruity between the options.

I think we can push this aside for now.

> Nimbus fails to compile for some reason.

Seems like an upstream problem.

**Comment by tupakkatapa** [comment]
> Sorry, meant `[ponkila-ephemeral-beta](https://github.com/ponkila/homestaking-infra/tree/main/nixosConfigurations/ponkila-ephemeral-beta)`. 

Me too, actually. I can only access the machine trough our VPN so I still need the configuration.

I just realized that I can get it from the sops file using the host's age key, even if my PGP key is not listed for the machine.. right? I still need to place it on the persistent drive because of the homestakeros though.

> Seems like an upstream problem.

Probably, I quickly searched the issues at [`status-im/nimbus-eth2`]( https://github.com/status-im/nimbus-eth2) and couldn't find anything exactly like this error. I might open an issue if I can't find anything related to it.

**Comment by jhvst** [comment]
>Me too, actually. I can only access the machine trough our VPN so I still need the configuration.
>
>I just realized that I can get it from the sops file using the host's age key, even if my PGP key is not listed for the machine.. right? I still need to place it on the persistent drive because of the homestakeros though.

`sops` keys are also always decrypted into `/run/secrets` folder. You can `cat` the secrets by running `sudo cat /run/secrets/wireguard/wg0` in this case.

> I might open an issue if I can't find anything related to it.

This is more likely something you should open at ethereum.nix first.

**Comment by jhvst** [comment]
Moreover, the first course of action for nimbus could be to update the nixobolus flake.

File: modules/homestakeros/default.nix [hunk 1]
```diff
@@ -223,7 +223,7 @@ in
               Type = "simple";
             };
 
-            script = ''${pkgs.go-ethereum}/bin/go-ethereum \
+            script = ''${pkgs.go-ethereum}/bin/geth \
```
Comment (jhvst) [inline modules/homestakeros/default.nix]: there is also this helper: `lib.getExe`. see, e.g: https://github.com/jhvst/nix-config/blob/main/home-manager/juuso.nix#L149

**Comment by tupakkatapa** [comment]
Tested every client ad hoc on our `ponkila-ephemeral-beta` machine. After a couple of fixes, everything seems to work fine and started to sync, noticing other client, except for Nimbus, which refuses to even compile.

I got the client commands conveniently using nix eval:

```shell
nix eval --raw .#nixosConfigurations.homestakeros.config.systemd.services.<clientName>.script
``````

<details>

<summary>Please note that this JSON data was loaded into homestakeros during this time. </summary>

```json
{
  "execution": {
    "erigon": {
      "enable": true,
      "dataDir": "/var/mnt/homestakeros/eth1/erigon",
      "jwtSecretFile": "/var/mnt/homestakeros/secrets/jwt.hex"
    },
    "geth": {
      "enable": true,
      "dataDir": "/var/mnt/homestakeros/eth1/geth",
      "jwtSecretFile": "/var/mnt/homestakeros/secrets/jwt.hex"
    },
    "nethermind": {
      "enable": true,
      "dataDir": "/var/mnt/homestakeros/eth1/nethermind",
      "jwtSecretFile": "/var/mnt/homestakeros/secrets/jwt.hex"
    },
    "besu": {
      "enable": true,
      "dataDir": "/var/mnt/homestakeros/eth1/besu",
      "jwtSecretFile": "/var/mnt/homestakeros/secrets/jwt.hex"
    }
  },
  "consensus": {
    "lighthouse": {
      "enable": true,
      "dataDir": "/var/mnt/homestakeros/eth2/lighthouse",
      "jwtSecretFile": "/var/mnt/homestakeros/secrets/jwt.hex",
    },
    "prysm": {
      "enable": true,
      "dataDir": "/var/mnt/homestakeros/eth2/prysm",
      "jwtSecretFile": "/var/mnt/homestakeros/secrets/jwt.hex",
    },
    "teku": {
      "enable": true,
      "dataDir": "/var/mnt/homestakeros/eth2/teku",
      "jwtSecretFile": "/var/mnt/homestakeros/secrets/jwt.hex",
    }
  }
}
```

</details>

### Executions clients (tested w/ lighthouse):

```shell
erigon \
    --datadir=/var/mnt/homestakeros/eth1/erigon \
    --chain mainnet \
    --authrpc.vhosts="*" \
    --authrpc.port 8551 \
    --authrpc.addr 127.0.0.1 \
    --authrpc.jwtsecret=/var/mnt/homestakeros/secrets/jwt.hex \
    --metrics

geth \
    --mainnet \
    --datadir /var/mnt/homestakeros/eth1/geth \
    --authrpc.vhosts "*" \
    --authrpc.port 8551 \
    --authrpc.addr 127.0.0.1 \
    --authrpc.jwtsecret /var/mnt/homestakeros/secrets/jwt.hex \
    --metrics

Nethermind.Runner \
    --config mainnet \
    --datadir /var/mnt/homestakeros/eth1/nethermind \
    --JsonRpc.EngineHost 127.0.0.1 \
    --JsonRpc.EnginePort 8551 \
    --JsonRpc.JwtSecretFile /var/mnt/homestakeros/secrets/jwt.hex \
    --Metrics.Enabled true

besu \
    --network=mainnet \
    --data-path=/var/mnt/homestakeros/eth1/besu \
    --engine-rpc-enabled=true \
    --engine-host-allowlist="*" \
    --engine-rpc-port=8551 \
    --rpc-http-enabled=true \
    --rpc-http-host=127.0.0.1 \
    --engine-jwt-secret=/var/mnt/homestakeros/secrets/jwt.hex \
    --metrics-enabled=true

```

### Consensus clients (tested w/ geth):

```shell
lighthouse bn \
    --datadir /var/mnt/homestakeros/eth2/lighthouse \
    --network mainnet \
    --http \
    --http-address 127.0.0.1 \
    --http-port 5052 \
    --http-allow-origin "*" \
    --execution-endpoint http://127.0.0.1:8551 \
    --execution-jwt /var/mnt/homestakeros/secrets/jwt.hex \
    --prune-payloads false \
    --metrics

teku \
    --data-base-path=/var/mnt/homestakeros/eth2/teku \
    --network=mainnet \
    --rest-api-enabled=true \
    --rest-api-port=5051 \
    --rest-api-interface=127.0.0.1 \
    --rest-api-host-allowlist="*" \
    --metrics-enabled=true \
    --ee-endpoint=http://127.0.0.1:8551 \
    --ee-jwt-secret-file=/var/mnt/homestakeros/secrets/jwt.hex

# prysm
beacon-chain \
    --accept-terms-of-use \
    --datadir /var/mnt/homestakeros/eth2/prysm \
    --mainnet \
    --grpc-gateway-host 127.0.0.1 \
    --grpc-gateway-port 3500 \
    --execution-endpoint http://127.0.0.1:8551 \
    --jwt-secret /var/mnt/homestakeros/secrets/jwt.hex
    
```

**Comment by tupakkatapa** [comment]
> Nimbus fails to compile for some reason.

Updating the flake did not have an effect. Converting to draft until Nimbus is updated (https://github.com/nix-community/ethereum.nix/pull/348). If the issue persists, I will investigate further.

**Comment by jhvst** [comment]
Is there an update to this? It might be useful to have this merged so that we can have a dropdown selector for clients on the frontend.

**Comment by tupakkatapa** [comment]
> Is there an update to this?

Nimbus has not been updated at ethereum.nix, should I remove Nimbus from this PR for the time being?

**Comment by jhvst** [comment]
Yea, more important to get this merged now

**Review by jhvst — APPROVED** [review]
I think the amount of refactoring in this PR makes this challenging to review properly, so what I suggest is that we merge this alongside #29 and then compare how would `nsq` outputs on homestaking-infra look like.

**Comment by selfuryon** [comment]
> > Nimbus fails to compile for some reason.
> 
> Updating the flake did not have an effect. Converting to draft until Nimbus is updated ([nix-community/ethereum.nix#348](https://github.com/nix-community/ethereum.nix/pull/348)). If the issue persists, I will investigate further.

We fixed the problem with nimbus, now it works, but need time to merge it :)

**Pull request merged**